src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceTypeOverrideService.kt

```diff
@@ -0,0 +1,124 @@
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceType.AP
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceType.AP_PSS
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceType.PSS
+import java.time.LocalDateTime
+
+@Service
+class LicenceTypeOverrideService(
+  private val licenceRepository: LicenceRepository,
+  private val auditEventRepository: AuditEventRepository,
+  private val staffRepository: StaffRepository,
+  private val licencePolicyService: LicencePolicyService,
+) {
+
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  @Transactional
+  fun changeType(licenceId: Long, newType: LicenceType, reason: String) {
+    val licence = licenceRepository.findById(licenceId).orElseThrow { EntityNotFoundException("$licenceId") }
+    val previousType = licence.typeCode
+
+    if (previousType == newType) {
+      log.info("Licence $licenceId already is of type: $newType")
+      return
+    }
+    log.info("Converting Licence $licenceId from $previousType to $newType")
+
+    getIncorrectDates(newType, licence).takeIf { it.isNotEmpty() }?.let { errors ->
+      throw DetailedValidationException(
+        title = "Incorrect dates for new licence type: $newType",
+        errors = errors.map { it.dateName to it.errorType.name },
+      )
+    }
+
+    val username = SecurityContextHolder.getContext().authentication.name
+    val staffMember = staffRepository.findByUsernameIgnoreCase(username)
+
+    licence.typeCode = newType
+
+    val (relevantConditions, conditionsToRemove) = licence.getAdditionalConditionsByRelevancy()
+    val standardConditions = licencePolicyService.getStandardConditionsForLicence(licence)
+
+    val licenceToPersist = licence.updateConditions(
+      updatedAdditionalConditions = relevantConditions,
+      updatedStandardConditions = standardConditions,
+      staffMember = staffMember,
+    )
+
+    licenceRepository.saveAndFlush(licenceToPersist)
+
+    auditEventRepository.saveAndFlush(
+      AuditEvent(
+        licenceId = licence.id,
+        detail = "ID ${licence.id} type ${licence.typeCode} status ${licence.statusCode} version ${licence.version}",
+        eventTime = LocalDateTime.now(),
+        eventType = AuditEventType.USER_EVENT,
+        username = staffMember?.username ?: SYSTEM_USER,
+        fullName = staffMember?.fullName ?: SYSTEM_USER,
+        summary = "Licence type overridden for ${licence.forename} ${licence.surname}: from $previousType to $newType: $reason",
+        changes = mutableMapOf(
+          "oldType" to previousType.name,
+          "newType" to newType.name,
+          "deletedAdditionalConditions" to conditionsToRemove.map {
+            mapOf(
+              "code" to it.conditionCode,
+              "version" to it.conditionVersion,
+              "category" to it.conditionCategory,
+              "text" to it.expandedConditionText,
```

Comment:
I was going to just persist the whole object but the backref on the condition entity to licence caused a stackoverflow